mutiny-core/src/nostr/mod.rs

```diff
@@ -0,0 +1,194 @@
+mod nip47;
+
+use crate::error::MutinyError;
+use crate::nodemanager::NodeManager;
+use crate::nostr::nip47::Nip47Request;
+use crate::storage::MutinyStorage;
+use crate::utils;
+use anyhow::anyhow;
+use bitcoin::hashes::hex::ToHex;
+use bitcoin::secp256k1::{PublicKey, Secp256k1};
+use bitcoin::util::bip32::{DerivationPath, ExtendedPrivKey};
+use lightning::util::logger::Logger;
+use lightning::{log_error, log_warn};
+use nostr::key::SecretKey;
+use nostr::prelude::{decrypt, encrypt};
+use nostr::{Event, EventBuilder, EventId, Filter, Keys, Kind, Tag, Timestamp};
+use nostr_sdk::Client;
+use serde_json::json;
+use std::str::FromStr;
+
+const MAX_ZAP_AMOUNT_SATS: u64 = 10_000;
+
+/// Manages Nostr keys and has different utilities for nostr specific things
+pub struct NostrManager {
+    /// Primary key used for nostr, this will be used for signing events
+    pub primary_key: Keys,
+    /// Client key used for Nostr Wallet Connect.
+    /// Mutiny will never use this key but it will be given to the client
+    /// in the connect URI.
+    nwc_client_key: Keys,
+    /// Server key used for Nostr Wallet Connect.
+    /// The nostr client will use this key to encrypt messages to the wallet.
+    /// Mutiny will use this key to decrypt messages from the nostr client.
+    nwc_server_key: Keys,
+    pub relays: Vec<String>,
+}
+
+impl NostrManager {
+    fn derive_nostr_key(xprivkey: ExtendedPrivKey, index: i32) -> Result<Keys, MutinyError> {
+        let context = Secp256k1::new();
+
+        let path = DerivationPath::from_str(&format!("m/44'/1237'/{index}'/0/0"))?;
+        let key = xprivkey.derive_priv(&context, &path)?;
+
+        // just converting to nostr secret key, unwrap is safe
+        let secret_key = SecretKey::from_slice(&key.private_key.secret_bytes()).unwrap();
+        Ok(Keys::new(secret_key))
+    }
+
+    /// Creates a new NostrManager
+    pub fn from_mnemonic(
+        xprivkey: ExtendedPrivKey,
+        relays: Vec<String>,
+    ) -> Result<Self, MutinyError> {
+        if relays.is_empty() {
+            return Err(MutinyError::Other(anyhow!("No relays provided")));
+        }
+
+        let primary_key = Self::derive_nostr_key(xprivkey, 0)?;
+        let nwc_client_key = Self::derive_nostr_key(xprivkey, 1)?;
+        let nwc_server_key = Self::derive_nostr_key(xprivkey, 2)?;
```

Comment:
I wonder if this should use the auth stuff at all? Otherwise these wont ever change or support multiple. What if they leak?

Reply:
Yeah I tried doing this originally but the lnurl derivation stuff is too different from the nostr derivation stuff. Maybe can unify one day but this was easier for now.